fe/fe-core/src/main/java/com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java

```diff
@@ -1864,17 +1866,24 @@ private void addRangePartitions(Database database, MaterializedView materialized
                     new SingleRangePartitionDesc(false, mvPartitionName, partitionKeyDesc, partitionProperties);
             partitionDescs.add(singleRangePartitionDesc);
         }
-        RangePartitionDesc rangePartitionDesc =
-                new RangePartitionDesc(materializedView.getPartitionColumnNames(), partitionDescs);
 
-        try {
-            GlobalStateMgr.getCurrentState().getLocalMetastore().addPartitions(
-                    database, materializedView.getName(),
-                    new AddPartitionClause(rangePartitionDesc, distributionDesc,
-                            partitionProperties, false));
-        } catch (Exception e) {
-            throw new DmlException("Expression add partition failed: %s, db: %s, table: %s", e, e.getMessage(),
-                    database.getFullName(), materializedView.getName());
+        // create partitions in small batch, to avoid create too many partitions at once
+        for (List<PartitionDesc> batch : ListUtils.partition(partitionDescs, CREATE_PARTITION_BATCH_SIZE)) {
+            RangePartitionDesc rangePartitionDesc =
+                    new RangePartitionDesc(materializedView.getPartitionColumnNames(), batch);
+            AddPartitionClause alterPartition = new AddPartitionClause(rangePartitionDesc, distributionDesc,
+                    partitionProperties, false);
+            try {
+                GlobalStateMgr.getCurrentState().getLocalMetastore().addPartitions(
+                        database, materializedView.getName(), alterPartition);
+            } catch (Exception e) {
+                throw new DmlException("Expression add partition failed: %s, db: %s, table: %s", e, e.getMessage(),
+                        database.getFullName(), materializedView.getName());
+            }
```

Comment:
Good catch! what about dropping partitions?

Reply:
1. drop partition is not a batch job but one-by-one
2. in normal case it would not drop a lot of partitions at once